webapp/src/utils/jira_issue_metadata.jsx

```diff
@@ -1,37 +1,39 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
+import memoize from 'memoize-one';
+
 // This is a replacement for the Array.flat() function which will be polyfilled by Babel
 // in our 5.16 release. Remove this and replace with .flat() then.
 const flatten = (arr) => {
     return arr.reduce((acc, val) => acc.concat(val), []);
 };
 
-export function getProjectValues(metadata) {
+export const getProjectValues = memoize((metadata) => {
```

Comment:
Are these functions really that expensive? How often do they get called with identical parameters? Also, is the equality comparison more expensive then the actual operation in some of these cases?

Reply:
> Are these functions really that expensive? How often do they get called with identical parameters?

Some are more expensive than others, especially when the project count is very high. They are called whenever the modal re-renders. It is most noticeable in the create issue modal. When you are typing text into one of the inputs, there is a noticeable lag between each keystroke.

The values being passed in, like `metadata` here, is data from the server after you select a project. So it is called with identical parameters very often.

> Also, is the equality comparison more expensive then the actual operation in some of these cases?

The equality comparison is a shallow compare, similar to how the selectors work from the `reselect` library, so it is very quick.

Reply:
@crspeller It's certainly not required to include these. The select box itself is not affected by it. I figured we should optimize the mid-render selectors, though if you feel strongly about not including them I'll take them out.

Also, running into `Too many nested callbacks (3). Maximum allowed is 2  max-nested-callbacks` on `getCustomFieldValuesForProjects()`, so I'll have to restructure the function a bit to compensate.

Reply:
@crspeller I decided to remove the memoization. It clutters the code, and the performance hit on 30,000 projects is noticeable, but isn't that bad. However, I'm wondering how it would affect a user with a low-end machine.